samza-core/src/main/java/org/apache/samza/execution/JobNodeConfigureGenerator.java

```diff
@@ -0,0 +1,340 @@
+import org.apache.samza.operators.spec.JoinOperatorSpec;
+import org.apache.samza.operators.spec.OperatorSpec;
+import org.apache.samza.operators.spec.StatefulOperatorSpec;
+import org.apache.samza.operators.spec.StoreDescriptor;
+import org.apache.samza.operators.spec.WindowOperatorSpec;
+import org.apache.samza.serializers.KVSerde;
+import org.apache.samza.serializers.Serde;
+import org.apache.samza.serializers.SerializableSerde;
+import org.apache.samza.table.TableConfigGenerator;
+import org.apache.samza.table.TableSpec;
+import org.apache.samza.util.MathUtil;
+import org.apache.samza.util.StreamUtil;
+import org.apache.samza.util.Util;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This class provides methods to generate configuration for a {@link JobNode}
+ */
+/* package private */ class JobNodeConfigureGenerator {
+
+  private static final Logger LOG = LoggerFactory.getLogger(JobNodeConfigureGenerator.class);
+
+  static final String CONFIG_JOB_PREFIX = "jobs.%s.";
+  static final String CONFIG_INTERNAL_EXECUTION_PLAN = "samza.internal.execution.plan";
+
+  JobNodeConfigureGenerator() {
+  }
+
+  static Config mergeJobConfig(Config originalConfig, Config generatedConfig) {
+    JobConfig jobConfig = new JobConfig(originalConfig);
+    String jobId = JobNode.createId(jobConfig.getName().get(), jobConfig.getJobId());
+    return new JobConfig(Util.rewriteConfig(extractScopedConfig(originalConfig, generatedConfig,
+        String.format(CONFIG_JOB_PREFIX, jobId))));
+  }
+
+  JobConfig generateJobConfig(JobNode jobNode, String executionPlanJson) {
+    Map<String, String> configs = new HashMap<>();
+    // set up job name and job ID
+    configs.put(JobConfig.JOB_NAME(), jobNode.getJobName());
+    configs.put(JobConfig.JOB_ID(), jobNode.getJobId());
+
+    Map<String, StreamEdge> inEdges = jobNode.getInEdges();
+    Map<String, StreamEdge> outEdges = jobNode.getOutEdges();
+    Collection<OperatorSpec> reachableOperators = jobNode.getReachableOperators();
+    boolean hasWindowOrJoin = reachableOperators.stream().anyMatch(op -> op.getOpCode() == OperatorSpec.OpCode.WINDOW
+        || op.getOpCode() == OperatorSpec.OpCode.JOIN);
+    List<StoreDescriptor> stores = getStoreDescriptors(reachableOperators);
+    Config config = jobNode.getConfig();
+
+    // check all inputs to the node for broadcast and input streams
+    final Set<String> inputs = new HashSet<>();
+    final Set<String> broadcasts = new HashSet<>();
+    for (StreamEdge inEdge : inEdges.values()) {
+      String formattedSystemStream = inEdge.getName();
+      if (inEdge.isBroadcast()) {
+        broadcasts.add(formattedSystemStream + "#0");
+      } else {
+        inputs.add(formattedSystemStream);
+      }
+    }
+
+    if (!broadcasts.isEmpty()) {
+      configureBroadcastInputs(configs, config, broadcasts);
+    }
+
+    // compute window and join operator intervals in this node
+    if (hasWindowOrJoin) {
```

Comment:
Minor: could the hasWindowOrJoin logic be encapsulated within the configureWindowAndJoinInterval?

Also, this method should be `configureWindowInterval`, since it's referring to configuring the task.window.ms.